pyam/core.py

```diff
@@ -510,14 +517,125 @@ def check_aggregate(self, variable, components=None, units=None,
+        exclude_on_fail: boolean, default False
+            flag scenarios failing validation as `exclude: True`
+        multiplier: number, default 1
+            factor when comparing variable and sum of components
+        kwargs: passed to `np.isclose()`
+        """
+        # default components to all variables one level below `variable`
+        var_df = self.filter(variable=variable, level=0)
+
+        if components is None:
+            components = var_df.filter(region=region, keep=False).regions()
+
+        if not len(components):
+            msg = (
+                '{} - cannot check regional aggregate because it has no '
+                'regional components'
+            )
+            logger().info(msg.format(variable))
+
+            return None
+
+        # filter and groupby data, use `pd.Series.align` for matching index
+        df_region, df_components = (
+            _aggregate_by_regions(var_df.data, region, units)
+            .align(_aggregate_by_regions(var_df.data, components, units))
+        )
+
+        # if it's a World comparison, we need to add in variables that might
+        # have been missed e.g. shipping and aviation
+        if region == "World":
+            df_components.index = df_components.index.droplevel(
+                "variable"
+            )
+            non_world_region = components[0]
+
+            for var_to_add in self.filter(variable="{}|*".format(variable)).variables():
+
+                var_has_regional_info = (
+                    (self.data.variable == var_to_add)
+                    & (self.data.region == non_world_region)
+                ).any()
+
+                if not var_has_regional_info:
+                    df_var_to_add = self.filter(region=region, variable=var_to_add).data.groupby(REGION_IDX).sum()['value']
+                    df_var_to_add.index = df_var_to_add.index.droplevel("variable")
+
+                    if len(df_var_to_add):
+                        df_components = df_components.add(df_var_to_add,
+                                                          fill_value=0)
+
+            df_components = pd.concat([df_components], keys=[variable],
+                                      names=['variable'])
+
+        # use `np.isclose` for checking match
+        diff = df_region[~np.isclose(df_region, multiplier * df_components,
+                                     **kwargs)]
+
+        if len(diff):
+            msg = (
+                '{} - {} of {} data points are not aggregates of regional '
+                'components'
+            )
+            logger().info(msg.format(variable, len(diff), len(df_region)))
+
+            if exclude_on_fail:
+                self._exclude_on_fail(diff.index.droplevel([2, 3]))
+
+            diff = pd.concat([diff], keys=[region], names=['region'])
+
+            return diff.unstack().rename_axis(None, axis=1)
+
+    def check_internal_consistency(self, **kwargs):
```

Comment:
from my experience with the SR1.5 database, there are so many variables that don't aggregate up "naturally", like `Final Energy|Gas` and `Final Energy|Transport`, that this function would be pretty much useless in general applications. I'd rather pass over a list of variables (with mapping to components were necessary) that should be checked rather than assuming a blanket "check everything".

Reply:
That may be the case for energy but for an emissions user (like us), these aggregations have to be the case. Maybe this should take a filter keyword so it could just look at variables that start with 'Emissions'?